File: app/src/main/java/tw/tonyyang/englishwords/util/EggTask.java
package tw.tonyyang.englishwords.util;

import android.app.ProgressDialog;
import android.content.Context;
import android.os.AsyncTask;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import tw.tonyyang.englishwords.R;

/**
 * Created by tonyyang on 2017/5/13.
 */

public abstract class EggTask<Params, Progress, Result> extends AsyncTask<Params, Progress, Result> {
    private static final Logger logger = LoggerFactory.getLogger(EggTask.class);

    protected ProgressDialog progress;

    protected Context context;

    protected boolean isShowProgressView = true;

    protected EggTask(Context context) {
        this.context = context;
        progress = new ProgressDialog(context);
        progress.setTitle(null);
        progress.setMessage(getProgressMessage());
        progress.setCancelable(false);
        progress.setCanceledOnTouchOutside(false);
    }

    protected String getProgressMessage() {
        return context.getString(R.string.loading_message);
    }

    @Override
    protected void onPreExecute() {
        if (isShowProgressView) {
            try {
                progress.show();
            } catch (Exception e) {
                logger.debug(String.valueOf(e));
            }
        }
    }

    @Override
    protected void onPostExecute(Result result) {
        try {
            progress.dismiss();
        } catch (IllegalArgumentException e) {
            logger.debug(String.valueOf(e));
        }
    }

    @Override
    protected void onCancelled() {
        super.onCancelled();
        try {
            progress.dismiss();
        } catch (IllegalArgumentException e) {
            logger.debug(String.valueOf(e));
        }
    }

    public void setShowProgressView(boolean isShowProgressView) {
        this.isShowProgressView = isShowProgressView;
    }
}

File: app/src/main/java/tw/tonyyang/englishwords/util/EggTask.kt
package tw.tonyyang.englishwords.util

import android.content.Context
import android.os.AsyncTask
import androidx.appcompat.app.AlertDialog
import org.slf4j.LoggerFactory
import tw.tonyyang.englishwords.R

abstract class EggTask<Params, Progress, Result> internal constructor(private val context: Context) : AsyncTask<Params, Progress, Result>() {

    private val progress: AlertDialog by lazy {
        UiUtils.getProgressDialog(context, context.getString(R.string.loading_message))
    }

    private var isShowProgressView = true

    fun setShowProgressView(isShowProgressView: Boolean) {
        this.isShowProgressView = isShowProgressView
    }

    override fun onPreExecute() {
        if (isShowProgressView) {
            try {
                progress.show()
            } catch (e: Exception) {
                logger.debug(e.toString())
            }
        }
    }

    override fun onPostExecute(result: Result) {
        try {
            progress.dismiss()
        } catch (e: IllegalArgumentException) {
            logger.debug(e.toString())
        }
    }

    override fun onCancelled() {
        super.onCancelled()
        try {
            progress.dismiss()
        } catch (e: IllegalArgumentException) {
            logger.debug(e.toString())
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(EggTask::class.java)
    }
}